processing/src/main/java/org/apache/druid/storage/remote/ChunkingStorageConnector.java

```diff
@@ -0,0 +1,204 @@
+
+  @Override
+  public InputStream read(String path) throws IOException
+  {
+    return buildInputStream(buildInputParams(path));
+  }
+
+  @Override
+  public InputStream readRange(String path, long from, long size)
+  {
+    return buildInputStream(buildInputParams(path, from, size));
+  }
+
+  public abstract ChunkingStorageConnectorParameters<T> buildInputParams(String path) throws IOException;
+
+  public abstract ChunkingStorageConnectorParameters<T> buildInputParams(String path, long from, long size);
+
+  private InputStream buildInputStream(ChunkingStorageConnectorParameters<T> params)
+  {
+    // Position from where the read needs to be resumed
+    final AtomicLong currentReadStartPosition = new AtomicLong(params.getStart());
+
+    // Final position, exclusive
+    long readEnd = params.getEnd();
+
+    AtomicBoolean isSequenceStreamClosed = new AtomicBoolean(false);
+
+    return new SequenceInputStream(
+
+        new Enumeration<InputStream>()
+        {
+          boolean initStream = false;
+
+          @Override
+          public boolean hasMoreElements()
+          {
+            // Checking if the stream was already closed. If it was, then don't iterate over the remaining chunks
+            // SequenceInputStream's close method closes all the chunk streams in its close. Since we're opening them
+            // lazily, we don't need to close them.
+            if (isSequenceStreamClosed.get()) {
+              return false;
+            }
+            // Don't stop until the whole object is downloaded
+            return currentReadStartPosition.get() < readEnd;
+          }
+
+          @Override
+          public InputStream nextElement()
+          {
+            if (!initStream) {
+              initStream = true;
+              return new NullInputStream();
+            }
+
+            File outFile = new File(
+                params.getTempDirSupplier().get().getAbsolutePath(),
+                UUID.randomUUID().toString()
+            );
+
+            long currentReadEndPosition = Math.min(
+                currentReadStartPosition.get() + chunkSizeBytes,
+                readEnd
+            );
+
+            try {
+              if (!outFile.createNewFile()) {
+                throw new IOE(
+                    StringUtils.format(
+                        "Could not create temporary file [%s] for copying [%s]",
+                        outFile.getAbsolutePath(),
+                        params.getCloudStoragePath()
+                    )
+                );
+              }
+
+              FileUtils.copyLarge(
+                  () -> new RetryingInputStream<>(
+                      params.getObjectSupplier().getObject(currentReadStartPosition.get(), currentReadEndPosition),
+                      params.getObjectOpenFunction(),
+                      params.getRetryCondition(),
+                      params.getMaxRetry()
+                  ),
+                  outFile,
+                  new byte[8 * 1024],
```

Comment:
I know this code was only moved, but could you add a comment on why these numbers are chosen?

Reply:
Sure